src/util/keyboardVisibility.ts

```diff
@@ -0,0 +1,74 @@
+import _ from 'lodash'
+import { editingActionCreator as editingAction } from '../actions/editing'
+import { isTouch } from '../browser'
+import { KEYBOARD_VISIBILITY_THRESHOLD } from '../constants'
+import * as selection from '../device/selection'
+import store from '../stores/app'
+
+let initialViewportHeight: number | null = window.visualViewport?.height || null
+let lastViewportHeight: number | null = window.visualViewport?.height || null
+let isKeyboardVisible: boolean = false
+
+/**
+ * Reset keyboard visibility tracker state.
+ */
+const reset = (): void => {
+  initialViewportHeight = null
+  lastViewportHeight = null
+  isKeyboardVisible = false
+}
+
+/**
+ * Handle viewport resize to detect keyboard visibility changes.
+ * Returns a throttled function to be used as an event handler.
+ */
+const handleKeyboardVisibility = _.throttle(() => {
+  // Only apply this for touch devices  where the issue occurs
+  if (!isTouch || !window.visualViewport) {
+    return
+  }
+
+  const visualViewport = window.visualViewport
+
+  // If we don't have initial height, set it now
+  if (initialViewportHeight === null) {
+    initialViewportHeight = visualViewport.height
+    lastViewportHeight = visualViewport.height
+    return
+  }
+
+  const currentHeight = visualViewport.height
+  const previousHeight = lastViewportHeight || initialViewportHeight
+
+  // Calculate height change ratio
+  const heightChangeRatio = Math.abs(currentHeight - previousHeight) / previousHeight
+
+  // Update last height
+  lastViewportHeight = currentHeight
```

Comment:
If you update `lastViewportHeight` after the threshold check, you may be able to eliminate `previousHeight`. 

Conceptually, `lastViewportHeight` and `previousHeight` appear to represent the same thing. Is that true, or am I missing something?